src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.CodeFixes/BinaryFormatterUnsafeDeserializeCodeFixer.cs

```diff
@@ -0,0 +1,100 @@
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.CodeFixes
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = "UA0012 CodeFix Provider")]
+    public sealed class BinaryFormatterUnsafeDeserializeCodeFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(BinaryFormatterUnsafeDeserializeAnalyzer.DiagnosticId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            if (root is null)
+            {
+                return;
+            }
+
+            var node = root.FindNode(context.Span, false, true);
+
+            if (node is null)
+            {
+                return;
+            }
+
+            var invocationExpression = node.Parent as InvocationExpressionSyntax;
+            if (invocationExpression is null)
+            {
+                return;
+            }
+
+            var lastArg = invocationExpression.ArgumentList.Arguments.Last();
+            if (!"null".Equals(lastArg.GetText().ToString(), System.StringComparison.Ordinal))
+            {
+                // UnsafeDeserialize accepts 2 parameters. This code fix only applies when the 2nd param is null
+                return;
+            }
+
+            // Register a code action that will invoke the fix.
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    CodeFixResources.BinaryFormatterUnsafeDeserializeTitle,
+                    cancellationToken => ReplaceUnsafeDeserializationAsync(context.Document, node, cancellationToken),
+                    nameof(CodeFixResources.BinaryFormatterUnsafeDeserializeTitle)),
+                context.Diagnostics);
+        }
+
+        private static async Task<Document> ReplaceUnsafeDeserializationAsync(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            if (node.Parent is null)
+            {
+                return document;
+            }
+
+            var project = document.Project;
+            var slnEditor = new SolutionEditor(project.Solution);
+
+            var docEditor = await slnEditor.GetDocumentEditorAsync(document.Id, cancellationToken).ConfigureAwait(false);
+            var docRoot = docEditor.OriginalRoot;
+
+            var replacementExpression = node.Parent.ToFullString();
```

Comment:
why handle this as a string? Seems like you should could do this at the syntax level

Reply:
Agreed, it seems like getting the syntax's expression as a MemberAccessExpressionSyntax and using WithName, and then using WithArgumentList or something like that would help to do this more directly.